packages/mobile/src/app/I18nGate.tsx

```diff
@@ -0,0 +1,40 @@
+import locales from 'locales'
+import { useEffect, useState } from 'react'
+import { findBestAvailableLanguage } from 'react-native-localize'
+import { useDispatch, useSelector } from 'react-redux'
+import { setLanguage } from 'src/app/actions'
+import { currentLanguageSelector } from 'src/app/reducers'
+import { DEFAULT_APP_LANGUAGE } from 'src/config'
+import { initI18n } from 'src/i18n'
+import Logger from 'src/utils/Logger'
+
+interface Props {
+  fallback: JSX.Element
+  children: JSX.Element
+}
+
+const I18nGate = ({ fallback, children }: Props) => {
+  const [isInitialised, setIsInitialised] = useState(false)
+  const dispatch = useDispatch()
+  const language = useSelector(currentLanguageSelector)
+  const bestLanguage = findBestAvailableLanguage(Object.keys(locales))?.languageTag
+
+  useEffect(() => {
+    const i18nInit = async () => {
+      try {
+        await initI18n(language || bestLanguage || DEFAULT_APP_LANGUAGE)
+        if (!language && bestLanguage) {
+          dispatch(setLanguage(bestLanguage))
+        }
+        setIsInitialised(true)
+      } catch (reason) {
+        Logger.error('i18n', 'Failed init i18n', reason)
```

Comment:
What is the app experience on error?  Is this app usable if `initI18n` fails?

Reply:
Ah, thanks for pointing this out! In this implementation the app will get stuck on the loading state indeed . we can either:
1. show the generic error screen here - unfortunately even the generic error screen requires us to use `t` to display the error, so if the `init` fails then we will see the translation keys
2. keep the current implementation, where we will allow the app to function normally. for a non english speaking user to see english dev keys, i'm not sure that's helpful at all...

i chose the error screen personally because i18n init is unlikely to fail, if it is failing then there are probably problems elsewhere so we can't guarantee that the app is working as expected. let me know if you think the second option is better, and i will change it :)

Reply:
sgtm!